File: gcpc2022/improvingit/submissions/accepted/lucas.py
#!/usr/bin/env python3
"""
Solves the problem in O(m * n).

author: Lucas Alber
"""
import sys

N, M = [int(x) for x in sys.stdin.readline().strip().split()]
costs = [0] + [1 << 63 for _ in range(N)]

for i in range(1, N + 1):
    cost, *values = [int(x) for x in sys.stdin.readline().strip().split()]
    for j in range(min(M, N - i + 1)):
        costs[i + j] = min(costs[i + j], costs[i - 1] + cost - values[j])

print(costs[N])

File: gcpc2022/improvingit/submissions/accepted/lucas.cpp
#include <iostream>
#include <vector>

using namespace std;

int main() {
    long long N;
    long long M;

    cin >> N >> M;

    vector<long long> costs(N + 1, 1ll << 62);
    costs[0] = 0;

    for (long long i = 1; i <= N + 1; i++) {
        long long cost;
        cin >> cost;

        for (long long j = 0; j < min(M, N - i + 1); j++) {
            long long value;
            cin >> value;

            costs[i + j] = min(costs[i + j], costs[i - 1] + cost - value);
        }
    }

    cout << costs[N] << "\n";
}